pyomo/core/expr/expr_common.py

```diff
@@ -82,6 +78,112 @@ class ExpressionType(enums.Enum):
     LOGICAL = 2
 
 
+class NUMERIC_ARG_TYPE(enums.IntEnum):
+    MUTABLE = -2
+    ASNUMERIC = -1
+    INVALID = 0
+    NATIVE = 1
+    NPV = 2
+    PARAM = 3
+    VAR = 4
+    MONOMIAL = 5
+    LINEAR = 6
+    SUM = 7
+    OTHER = 8
+
+
+class RELATIONAL_ARG_TYPE(enums.IntEnum, metaclass=enums.ExtendedEnumType):
+    __base_enum__ = NUMERIC_ARG_TYPE
+
+    INEQUALITY = 100
+    INVALID_RELATIONAL = 101
+
+
+def _invalid(*args):
+    return NotImplemented
+
+
+def _recast_mutable(expr):
+    expr.make_immutable()
+    if expr._nargs > 1:
+        return expr
+    elif not expr._nargs:
+        return 0
+    else:
+        return expr._args_[0]
+
+
+def _unary_op_dispatcher_type_mapping(dispatcher, updates, TYPES=NUMERIC_ARG_TYPE):
+    #
+    # Special case (wrapping) operators
+    #
+    def _asnumeric(a):
+        a = a.as_numeric()
+        return dispatcher[a.__class__](a)
+
+    def _mutable(a):
+        a = _recast_mutable(a)
+        return dispatcher[a.__class__](a)
+
+    mapping = {
+        TYPES.ASNUMERIC: _asnumeric,
+        TYPES.MUTABLE: _mutable,
+        TYPES.INVALID: _invalid,
+    }
+
+    mapping.update(updates)
+    return mapping
+
+
+def _binary_op_dispatcher_type_mapping(dispatcher, updates, TYPES=NUMERIC_ARG_TYPE):
+    #
+    # Special case (wrapping) operators
+    #
+    def _any_asnumeric(a, b):
+        b = b.as_numeric()
+        return dispatcher[a.__class__, b.__class__](a, b)
+
+    def _asnumeric_any(a, b):
+        a = a.as_numeric()
+        return dispatcher[a.__class__, b.__class__](a, b)
+
+    def _asnumeric_asnumeric(a, b):
+        a = a.as_numeric()
+        b = b.as_numeric()
+        return dispatcher[a.__class__, b.__class__](a, b)
+
+    def _any_mutable(a, b):
+        b = _recast_mutable(b)
+        return dispatcher[a.__class__, b.__class__](a, b)
+
+    def _mutable_any(a, b):
+        a = _recast_mutable(a)
+        return dispatcher[a.__class__, b.__class__](a, b)
+
+    def _mutable_mutable(a, b):
+        if a is b:
+            a = b = _recast_mutable(a)
+        else:
+            a = _recast_mutable(a)
+            b = _recast_mutable(b)
```

Comment:
Mostly just out of curiosity (because knowing you, you checked): is the `if` really so much faster than the `else` that it's worth checking and only recasting once?

Reply:
The real issue is that if `a is b`, we would be calling `_recast_mutable` on the same object twice.  This is a partially inplace operation (the first step is to switch the object from a mutable sum-like expression into a "normal" sum-like expression).  If you call `_recast_mutable` on the same object twice you will get an exception (an `AttributeError`, because once the object is not longable mutable, it no longer has the `make_immutable()` method).

Reply:
Oh, I see. Could you maybe add a comment to that effect, because I think only future-you will remember that!